File: packages/maplibre_android/android/src/main/kotlin/com/github/josxha/maplibre/MapLibrePlugin.kt
package com.github.josxha.maplibre

// if imports can't resolve:
// - remove all .idea/ folders
// - open example/android/build.gradle.kts as project
// - sync project to download dependencies

import android.app.Activity
import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import org.maplibre.android.location.permissions.PermissionsManager

/** MapLibrePlugin */
class MapLibrePlugin :
    FlutterPlugin,
    ActivityAware,
    PluginRegistry.RequestPermissionsResultListener {
    private var permissionsManager: PermissionsManager? = null
    private var requestHeadersChannel: MethodChannel? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        installHostScopedRequestHeadersInterceptor()
        requestHeadersChannel =
            MethodChannel(binding.binaryMessenger, requestHeadersChannelName).apply {
                setMethodCallHandler { call, result ->
                    val host = call.argument<String>("host")
                    if (host == null) {
                        result.error("invalid_arguments", "Missing host.", null)
                        return@setMethodCallHandler
                    }
                    when (call.method) {
                        "setRequestHeaders" -> {
                            val rawHeaders = call.argument<Map<*, *>>("headers")
                            val headers =
                                rawHeaders?.entries?.associate { entry ->
                                    entry.key as String to entry.value as String
                                }
                            if (headers == null) {
                                result.error(
                                    "invalid_arguments",
                                    "Missing headers.",
                                    null,
                                )
                            } else {
                                HostScopedRequestHeaders.replace(host, headers)
                                result.success(null)
                            }
                        }
                        "clearRequestHeaders" -> {
                            HostScopedRequestHeaders.clear(host)
                            result.success(null)
                        }
                        else -> result.notImplemented()
                    }
                }
            }
        binding
            .platformViewRegistry
            .registerViewFactory(
                "plugins.flutter.io/maplibre",
                MapLibreMapFactory(),
            )
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        requestHeadersChannel?.setMethodCallHandler(null)
        requestHeadersChannel = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray,
    ): Boolean {
        permissionsManager?.onRequestPermissionsResult(
            requestCode,
            permissions,
            grantResults,
        )
        return true
    }

    private companion object {
        const val requestHeadersChannelName =
            "plugins.flutter.io/maplibre/request_headers"
    }
}

class MapLibreMapFactory : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(
        context: Context,
        viewId: Int,
        args: Any?,
    ): PlatformView = MapLibreRegistry.flutterApi!!.createPlatformView(viewId)
}

File: packages/maplibre_ios/ios/maplibre_ios/Sources/maplibre_ios/MapLibrePlugin.swift
import Flutter
import MapLibre
import UIKit

public class MapLibrePlugin: NSObject, FlutterPlugin {
    private static let requestHeaders = HostScopedRequestHeaders()

    public static func register(with registrar: FlutterPluginRegistrar) {
        requestHeaders.install()
        let channel = FlutterMethodChannel(
            name: "plugins.flutter.io/maplibre/request_headers",
            binaryMessenger: registrar.messenger()
        )
        let instance = MapLibrePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)

        // register MapLibre view factory
        let factory = MapLibreViewFactory(withRegistrar: registrar)
        registrar.register(factory, withId: "plugins.flutter.io/maplibre")
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard
            let arguments = call.arguments as? [String: Any],
            let host = arguments["host"] as? String
        else {
            result(
                FlutterError(
                    code: "invalid_arguments",
                    message: "Missing host.",
                    details: nil
                )
            )
            return
        }

        switch call.method {
        case "setRequestHeaders":
            guard let headers = arguments["headers"] as? [String: String] else {
                result(
                    FlutterError(
                        code: "invalid_arguments",
                        message: "Missing headers.",
                        details: nil
                    )
                )
                return
            }
            Self.requestHeaders.replace(host: host, headers: headers)
            result(nil)
        case "clearRequestHeaders":
            Self.requestHeaders.clear(host: host)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
